diff --git Configure.pl Configure.pl
index 76e3e67..7560fcf 100644
--- Configure.pl
+++ Configure.pl
@@ -6,9 +6,13 @@
 use 5.008;
 use strict;
 use warnings;
-use lib 'lib';
+use FindBin qw($Bin);
+use File::Spec;
+use lib File::Spec->catdir($Bin, "lib");
 
 use Parrot::Configure;
+use Cwd qw(realpath);
+use File::Basename;
 use Parrot::Configure::Options qw( process_options );
 use Parrot::Configure::Options::Test;
 use Parrot::Configure::Options::Test::Prepare qw(
@@ -63,13 +67,18 @@ $conf->add_steps( @{ $steps_list_ref } );
 
 # from Parrot::Configure::Data
 $conf->options->set( %{$args} );
+# handles out of directory builds
+my $source_dir = dirname(realpath($0));
+my $build_dir = $ENV{PWD};
+$conf->options->set(build_dir => $build_dir);
+$conf->options->set(source_dir => $source_dir);
 # save the command-line for make reconfig
 $conf->data->set(configure_args => @ARGV ? '"'.join("\" \"", map {qq($_)} @ARGV).'"'
                                          : '');
 
 # Log files created by Configure.pl in MANIFEST.configure.generated
 $conf->{active_configuration} = 1;
-unlink 'MANIFEST.configure.generated';
+unlink File::Spec->catfile($build_dir, 'MANIFEST.configure.generated');
 
 # Run the actual steps from Parrot::Configure
 $conf->runsteps or exit(1);
diff --git config/init/headers.pm config/init/headers.pm
index b7c0128..9c83352 100644
--- config/init/headers.pm
+++ config/init/headers.pm
@@ -16,6 +16,7 @@ package init::headers;
 
 use strict;
 use warnings;
+use File::Spec;
 
 
 use base qw(Parrot::Configure::Step);
@@ -34,7 +35,7 @@ sub _init {
 sub runstep {
     my ( $self, $conf ) = @_;
 
-    my $inc = 'include/parrot';
+    my $inc = File::Spec->catdir($conf->options->get('source_dir'), 'include/parrot');
 
     my @headers = (
         sort
diff --git config/init/hints.pm config/init/hints.pm
index c4ed805..9b79150 100644
--- config/init/hints.pm
+++ config/init/hints.pm
@@ -40,7 +40,7 @@ sub runstep {
     $osname = 'linux' if ($osname eq 'gnukfreebsd');
 
     my $hints_file_name = $conf->options->get('hintsfile') || $osname ;
-    $hints_file = catfile('config', 'init', 'hints', "$hints_file_name.pm");
+    $hints_file = catfile($conf->options->get('source_dir'), 'config', 'init', 'hints', "$hints_file_name.pm");
 
     if ( -f $hints_file ) {
         my $hints_pkg = "init::hints::" . $hints_file_name;
diff --git config/init/manifest.pm config/init/manifest.pm
index e402cef..2096f9c 100644
--- config/init/manifest.pm
+++ config/init/manifest.pm
@@ -21,6 +21,7 @@ use base qw(Parrot::Configure::Step);
 use Parrot::Configure::Step;
 use Parrot::Configure::Utils ':gen';
 use ExtUtils::Manifest qw(manicheck);
+use Cwd qw(chdir);
 
 
 sub _init {
@@ -40,7 +41,11 @@ sub runstep {
         return 1;
     }
 
+    chdir($conf->options->get('source_dir'));
     my @missing = ExtUtils::Manifest::manicheck();
+    # The following would avoid a chdir but doesn't work:
+    # $ExtUtils::Manifest::MANIFEST = File::Spec->catfile($conf->options->get('source_dir'), 'MANIFEST');
+    # my @missing = ExtUtils::Manifest::_check_files($conf->options->get('source_dir'));
 
     if (@missing) {
         print <<"END";
diff --git config/inter/progs.pm config/inter/progs.pm
index d0b3347..4ddd1d3 100644
--- config/inter/progs.pm
+++ config/inter/progs.pm
@@ -21,6 +21,7 @@ use warnings;
 use base qw(Parrot::Configure::Step);
 
 use Parrot::Configure::Utils ':inter', ':auto';
+use File::Spec;
 
 
 sub _init {
@@ -159,8 +160,8 @@ sub _set_debug_and_warn {
 
 sub test_compiler {
     my ($conf, $cc) = @_;
-
-    open( my $out_fh, '>', "test_$$.c" )
+    my $build_dir = $conf->options->get('build_dir');
+    open( my $out_fh, '>', File::Spec->catfile($build_dir, "test_$$.c") )
         or die "Unable to open 'test_$$.c': $@\n";
     print {$out_fh} <<END_C;
 int main() {
diff --git lib/Parrot/BuildUtil.pm lib/Parrot/BuildUtil.pm
index f2dfb51..cd51844 100644
--- lib/Parrot/BuildUtil.pm
+++ lib/Parrot/BuildUtil.pm
@@ -40,17 +40,16 @@ sub parrot_version {
     if ( defined $parrot_version ) {
         return wantarray ? @parrot_version : $parrot_version;
     }
-
     # Obtain the official version number from the VERSION file.
-    if (-e 'VERSION') {
-        open my $VERSION, '<', 'VERSION' or die 'Could not open VERSION file!';
+    if (-e "VERSION") {
+        open my $VERSION, '<', "VERSION" or die "Could not open VERSION file!";
         chomp( $parrot_version = <$VERSION> );
         close $VERSION or die $!;
     }
     else { # we're in an installed copy of Parrot
         my $path = shift;
         $path = '' unless $path;
-        open my $VERSION, '<', "$path/VERSION" or die 'Could not open VERSION file!';
+        open my $VERSION, '<', "$path/VERSION" or die "Could not open $path/VERSION file!";
         chomp( $parrot_version = <$VERSION> );
         close $VERSION or die $!;
     }
diff --git lib/Parrot/Configure.pm lib/Parrot/Configure.pm
index 819bd23..3b20687 100644
--- lib/Parrot/Configure.pm
+++ lib/Parrot/Configure.pm
@@ -38,8 +38,9 @@ This module accepts no arguments to its C<import> method and exports no
 I<symbols>.
 
 =cut
-
-use lib qw(config);
+use FindBin qw($Bin);
+use File::Spec;
+use lib File::Spec->catdir($Bin, 'config');
 use Carp qw(carp);
 use Storable qw(2.12 nstore retrieve nfreeze thaw);
 use Parrot::Configure::Data;
@@ -375,6 +376,11 @@ sub _run_this_step {
 
     my $step_name   = $args->{task}->step;
 
+    my @dir = File::Spec->splitdir($Bin);
+    if ($dir[-2] eq 't') {
+        eval "use lib qw(" . File::Spec->catdir(@dir[0..$#dir - 2], 'config') . ");";
+    }
+
     eval "use $step_name;"; ## no critic (BuiltinFunctions::ProhibitStringyEval)
     die $@ if $@;
 
diff --git lib/Parrot/Configure/Compiler.pm lib/Parrot/Configure/Compiler.pm
index 8609775..6c181e9 100644
--- lib/Parrot/Configure/Compiler.pm
+++ lib/Parrot/Configure/Compiler.pm
@@ -89,6 +89,7 @@ Calls the compiler and linker on F<test_$$.c>.
 sub cc_build {
     my $conf = shift;
     my ( $cc_args, $link_args ) = @_;
+    my $build_dir = $conf->options->get('build_dir');
 
     $cc_args   = '' unless defined $cc_args;
     $link_args = '' unless defined $link_args;
@@ -101,16 +102,21 @@ sub cc_build {
     # unique test file name for parallel builds
     my $test            = 'test_' . $$;
     my $compile_command = _build_compile_command( $cc, $ccflags, $cc_args );
-    my $compile_result  = _run_command( $compile_command, "$test.cco", "$test.cco", $verbose );
+    my $actual_file = File::Spec->catfile($build_dir, $test);
+    $compile_command =~ s/$test/$actual_file/;
+    $compile_command .= " $cc_exe_out$actual_file$o";
+    my $output = "$actual_file.cco";
+    my $compile_result  = _run_command( $compile_command, $output, $output, $verbose );
 
     if ($compile_result) {
-        confess "C compiler failed (see $test.cco)";
+        confess "C compiler failed (see $actual_file.cco)";
         return $compile_result;
     }
 
+    $output = "$actual_file.ldo";
     my $link_result =
-        _run_command( "$link $linkflags $test$o $link_args ${cc_exe_out}${test}${exe}  $libs",
-        "$test.ldo", "$test.ldo", $verbose )
+        _run_command( "$link $linkflags $actual_file$o $link_args ${cc_exe_out}${actual_file}${exe}  $libs",
+        $output, $output, $verbose )
         and confess "Linker failed (see $test.ldo)";
     if ($link_result) {
         return $link_result;
@@ -133,6 +139,7 @@ sub cc_run {
     my $verbose  = $conf->options->get('verbose');
     my $test     = 'test_' . $$;
     my $test_exe = ".${slash}${test}${exe}";
+    my $build_dir = $conf->options->get('build_dir');
 
     my $run_error;
     if ( defined( $_[0] ) && length( $_[0] ) ) {
@@ -140,10 +147,10 @@ sub cc_run {
         $run_error = _run_command( "$test_exe @_", "./$test.out", undef, $verbose );
     }
     else {
-        $run_error = _run_command( $test_exe, "./$test.out", undef, $verbose );
+        $run_error = _run_command( File::Spec->catfile($build_dir, $test_exe), File::Spec->catfile($build_dir, "$test.out"), undef, $verbose );
     }
 
-    my $output = _slurp("./$test.out");
+    my $output = _slurp(File::Spec->catfile($build_dir, "$test.out"));
 
     return $output;
 }
@@ -333,6 +340,8 @@ syntax works ok.
 
 sub genfile {
     my $conf = shift;
+    my $source_dir = $conf->options->get('source_dir');
+    my $build_dir = $conf->options->get('build_dir');
     my ( $source, $target, %options ) = @_;
 
     my $calling_sub = (caller(1))[3] || q{};
@@ -340,8 +349,8 @@ sub genfile {
         $conf->append_configure_log($target);
     }
 
-    open my $in,  '<', $source       or die "Can't open $source: $!";
-    open my $out, '>', "$target.tmp" or die "Can't open $target.tmp: $!";
+    open my $in,  '<', ($source_dir ? File::Spec->catfile($source_dir,$source) : $source)       or die "Can't open $source: $!";
+    open my $out, '>', ($build_dir ? File::Spec->catfile($build_dir, "$target.tmp") : "$target.tmp")   or die "Can't open $target.tmp: $!";
 
     if ( !exists $options{file_type}) {
         if ( $target =~ m/makefile$/i || $target =~ m/\.mak/) {
@@ -552,7 +561,7 @@ sub genfile {
     close($in)  or die "Can't close $source: $!";
     close($out) or die "Can't close $target: $!";
 
-    move_if_diff( "$target.tmp", $target, $options{ignore_pattern} );
+    move_if_diff( File::Spec->catfile($build_dir, "$target.tmp"), File::Spec->catfile($build_dir, $target), $options{ignore_pattern} );
 }
 
 # Return the next subexpression from the expression in $_[0]
@@ -673,8 +682,9 @@ sub cond_eval {
 sub append_configure_log {
     my $conf = shift;
     my $target = shift;
+    my $build_dir = $conf->options->get('build_dir');
     if ( $conf->{active_configuration} ) {
-        my $generated_log = 'MANIFEST.configure.generated';
+        my $generated_log = File::Spec->catfile($build_dir, 'MANIFEST.configure.generated');
         open my $GEN, '>>', $generated_log
             or die "Can't open $generated_log for appending: $!";
         print $GEN "$target\n";
diff --git lib/Parrot/Configure/Options/Conf.pm lib/Parrot/Configure/Options/Conf.pm
index c3a58f9..02328d0 100644
--- lib/Parrot/Configure/Options/Conf.pm
+++ lib/Parrot/Configure/Options/Conf.pm
@@ -17,7 +17,7 @@ use Parrot::BuildUtil ();
 use FindBin qw($Bin);
 
 our $script         = q{Configure.pl};
-our $parrot_version = Parrot::BuildUtil::parrot_version("$Bin/../../");
+our $parrot_version = Parrot::BuildUtil::parrot_version($Bin);
 our $svnid          = '$Id$';
 
 sub print_version {
diff --git lib/Parrot/Configure/Options/Test/Prepare.pm lib/Parrot/Configure/Options/Test/Prepare.pm
index 7f73011..7107b1d 100644
--- lib/Parrot/Configure/Options/Test/Prepare.pm
+++ lib/Parrot/Configure/Options/Test/Prepare.pm
@@ -6,6 +6,8 @@ use warnings;
 use Carp;
 use Data::Dumper;$Data::Dumper::Indent=1;
 use File::Find;
+use FindBin qw($Bin);
+use File::Spec;
 #use Scalar::Util qw( looks_like_number );
 use base qw( Exporter );
 our @EXPORT_OK = qw(
@@ -14,11 +16,15 @@ our @EXPORT_OK = qw(
 );
 use lib qw(lib);
 use Parrot::Configure::Step::List qw( get_steps_list );
-
-my $config_dir = q{t/configure};
+my $path = $Bin;
+my @dirs = File::Spec->splitdir($path);
+if ($dirs[-2] eq 't' && $dirs[-1] eq 'configure') {
+  $path = '.';
+}
+my $config_dir = File::Spec->catdir($path, 't', 'configure');
 my @framework_tests = _get_framework_tests($config_dir);
 
-my $steps_dir = q{t/steps};
+my $steps_dir = File::Spec->catdir($path, 't', 'steps');
 my ( $steps_tests_simple_ref, $steps_tests_complex_ref )  =
     _find_steps_tests($steps_dir);
 my @steps_expected = get_steps_list();
@@ -35,8 +41,8 @@ sub get_preconfiguration_tests {
 sub get_postconfiguration_tests {
     my @postconfiguration_tests = (
         @steps_tests,
-        glob("t/postconfigure/*.t"),
-        glob("t/pharness/*.t"),
+        glob(File::Spec->catfile($path, 't', 'postconfigure', '*.t')),
+        glob(File::Spec->catfile($path, 't', 'pharness', '*.t')),
     );
     return @postconfiguration_tests;
 };
